### PR TITLE
#33171: Invalid link for readonly fields in django admin with missing…

### PR DESCRIPTION
**Pull request by GitRon**
… permissions

#### Trac ticket number

[ticket-33171](https://code.djangoproject.com/ticket/33171#ticket)

#### Branch description

Having the permission to see model A but not FK-related model B will display a link to model B. After clicking on this link, you'll end up on a 403 page. This is at least bad UX and probably a bug.

Thx to @sarahboyce in the [Django forum](https://forum.djangoproject.com/t/admin-403-link-with-missing-view-permission/33077/5) for providing this solution.

#### Checklist
- [x] This PR targets the `main` branch. 
- [ ] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [ ] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

**Comment by GitRon** [comment]
Hi @sarahboyce 

it seems your idea is flying. What are the next steps for me? How can I unit-test this? Are there some ressources I should be reading?

Best and thx so much for the input!  
Ronny

File: django/contrib/admin/templates/admin/includes/fieldset.html [hunk 1]
```diff
@@ -20,7 +20,13 @@
                             {% else %}
                                 {{ field.label_tag }}
                                 {% if field.is_readonly %}
-                                    <div class="readonly">{{ field.contents }}</div>
+                                    <p>
```
Comment (sarahboyce) [inline django/contrib/admin/templates/admin/includes/fieldset.html:23]: Look like this needs to be `<div class="readonly">` for some existing tests

**Review by sarahboyce — COMMENTED** [review]
> How can I unit-test this? Are there some resources I should be reading?

These are our docs on tests: https://docs.djangoproject.com/en/5.1/internals/contributing/writing-code/unit-tests/
You will want a "regression" test (so something that was failing before this change where the link was available even though you didn't have permission to view it)
You will also want full coverage of the new logic (to confirm all bits are required)

(edited for the correct link)

File: django/contrib/admin/templates/admin/includes/fieldset.html (hunk 1 above)
Comment (GitRon) [reply django/contrib/admin/templates/admin/includes/fieldset.html]: true, fixed it.

File: django/contrib/admin/templates/admin/includes/fieldset.html [hunk 2]
```diff
@@ -20,7 +20,13 @@
...
                                 {{ field.label_tag }}
                                 {% if field.is_readonly %}
-                                    <div class="readonly">{{ field.contents }}</div>
+                                    <div class="readonly">
+                                      {% if field.required_perm and field.required_perm not in perms %}
+                                        {{ field.contents|striptags }}
+                                      {% else %}
+                                        {{ field.contents }}
+                                      {% endif %}
+                                    </div>
```
Comment (sarahboyce) [inline django/contrib/admin/templates/admin/includes/fieldset.html:29]:
```suggestion
                                    <div class="readonly">{% if field.required_perm and field.required_perm not in perms %}{{ field.contents|striptags }}{% else %}{{ field.contents }}{% endif %}</div>
```
The tests appear to be new line sensitive

**Comment by GitRon** [comment]
Hi @sarahboyce! I have a bit of a hard time to write the tests. 

It's kind of hard deciding what to test where, there are SO many tests 😅 Secondly, I don't manage to get the runtests script to work properly, even though reading [the docs](https://docs.djangoproject.com/en/dev/intro/contributing/#writing-some-tests-for-your-ticket).

For some reason I can't see the logs from the broken tests here in GitHub... am I missing something?

But I tried to create a test, could you give me a hint if I'm on the right track here? I know that it's not done yet (I have to at least assign some permissions) but before adding more complexity, I'd like to know that it's ok in general.

Sorry to bother you with noobie questions 🙈

**Comment by GitRon** [comment]
@sarahboyce Finally managed to get the test running. What would you say we need to finish this PR?

**Comment by nessita** [comment]
Hello @GitRon! Thank you for the work done in this PR. :star2: 

Sarah and I were discussing this issue, and she made some suggestions and provided a resolution in the ticket-33171. Given that, we are keen to close this PR. If you wish to continue experimenting, let us know!

**Comment by GitRon** [comment]
Hi @nessita! Thx for the heads-up and @sarahboyce for all the work you've put into this!

**Pull request closed without merging**